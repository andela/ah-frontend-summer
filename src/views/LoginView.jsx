import React, { Component } from "react";
import Login from "../components/Login";
import { connect } from "react-redux";
import { loginFetch } from "../store/actions/async/login";
import { removeLoginError } from "../store/actions/sync/login";

export class LoginView extends Component {
    constructor(props) {
        super(props);
        this.state = {
        email: "",
        password: ""
        };
    }

    onSubmitHandler = e => {
        e.preventDefault();
        const { email, password } = this.state;
        const data = { email, password };
        const url =
        "https://ah-backend-summer-staging.herokuapp.com/api/v1/users/login";
        const payload = {
        data,
        history: this.props.history,
        url
        };
        this.props.login(payload);
    };

    onChangeHandler = e => this.setState({ [e.target.name]: e.target.value });

    onDismissHandler = () => this.props.dismissLoginError();

    render() {
        const loginProps = {
        email: this.state.email,
        password: this.state.password,
        onDismissHandler: this.onDismissHandler,
        loginError: this.props.loginError,
        onSubmitHandler: this.onSubmitHandler,
        onChangeHandler: this.onChangeHandler
        };
        return <Login {...loginProps} />;
    }
}

const mapDispatchToProps = dispatch => {
    return {
        login: payload => dispatch(loginFetch(payload)),
        dismissLoginError: () => dispatch(removeLoginError())
    };
};

const mapStateToProps = state => {
    return {
        loginError: state.login.loginError
    };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginView);
