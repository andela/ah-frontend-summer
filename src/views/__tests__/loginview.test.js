import React from "react";
import { shallow } from "enzyme";
import { LoginView, mapDispatchToProps, mapStateToProps } from "../LoginView";

describe("login view", () => {
    const mockLoginFetch = jest.fn();
    const mockRemoveLoginError = jest.fn();
    const mockFacebookLogin = jest.fn();
    const mockGoogleLogin = jest.fn();
    const dispatch = jest.fn();

    const loginViewProps = {
        login: mockLoginFetch,
        dismissLoginError: mockRemoveLoginError,
        loginError: "",
        loading: false,
        facebookLogin: mockFacebookLogin,
        googleLogin: mockGoogleLogin,
        history: {}
    };

    const loginView = shallow(<LoginView {...loginViewProps} />);
    const loginViewInstance = loginView.instance();

    afterEach(() => {
        loginView.setState({ email: "", password: "" });
    });

    it("renders loginView component correctly", () => {
        expect(loginView).toMatchSnapshot();
    });

    it("onChangeHandler updates email in state when an event with value of email is passed", () => {
        loginViewInstance.onChangeHandler({
            target: { value: "okay@email.com", name: "email" }
        });
        expect(loginView.state().email).toBe("okay@email.com");
    });

    it("onChangeHandler updates password in state when an event with value of password is passed", () => {
        loginViewInstance.onChangeHandler({
            target: { value: "okay", name: "password" }
        });
        expect(loginView.state().password).toBe("okay");
    });

    it("onSubmitHandler dispatches loginFetch action creator when called", () => {
        loginViewInstance.onSubmitHandler({ preventDefault: () => null });
        expect(mockLoginFetch).toHaveBeenCalled();
    });

    it("onDismissHandler dispatches removeLoginError action creator to remove login error message", () => {
        loginViewInstance.onDismissHandler();
        expect(mockRemoveLoginError).toHaveBeenCalled();
    });
    it("responseFacebook dispatches loginFetch action creator when called", () => {
        loginViewInstance.responseFacebook({ accessToken: 'jjajisi3' });
        expect(mockFacebookLogin).toHaveBeenCalled();
    });
    it("responseGoogle dispatches loginFetch action creator when called", () => {
        loginViewInstance.responseGoogle({ tokenId: 'jjajisi3' });
        expect(mockGoogleLogin).toHaveBeenCalled();
    });

    it("should map dispatch to props", () => {
        mapDispatchToProps(dispatch).facebookLogin();
        mapDispatchToProps(dispatch).googleLogin();
        mapDispatchToProps(dispatch).login();
        mapDispatchToProps(dispatch).dismissLoginError();
        expect(dispatch.mock.calls.length).toBe(4);
    });

    it("should map state to props", () => {
        const login = { loginError: "am an error", loading: "" };
        expect(mapStateToProps({ login })).toEqual({ loginError: "am an error", loading: "" });
    });
});
