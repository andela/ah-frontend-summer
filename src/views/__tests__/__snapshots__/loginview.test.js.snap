// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login view renders loginView component correctly 1`] = `
<Login
  email=""
  onChangeHandler={[Function]}
  onDismissHandler={[Function]}
  onSubmitHandler={[Function]}
  password=""
/>
`;
